File: peer_ios/peerapp/PeerClient.swift
//
//  PeerClient.swift
//  peerapp
//
//  Created by emblock on 12/17/25.
//
import Foundation
import Network

class PeerClient {
    private let serverIp: String
    private let serverPort: UInt16
    private let onMessageReceived: (String) -> Void
    
    private var connection: NWConnection?
    private let queue = DispatchQueue(label: "PeerClientQueue")

    init(serverIp: String, serverPort: UInt16 = 8888, onMessageReceived: @escaping (String) -> Void) {
        self.serverIp = serverIp
        self.serverPort = serverPort
        self.onMessageReceived = onMessageReceived
    }

    func start() {
        guard connection == nil else {
            print("PeerClient is already running.")
            return
        }

        let host = NWEndpoint.Host(serverIp)
        let port = NWEndpoint.Port(rawValue: serverPort)!
        
        connection = NWConnection(host: host, port: port, using: .udp)
        
        connection?.stateUpdateHandler = { newState in
            switch newState {
            case .ready:
                print("PeerClient started. Server: \(self.serverIp):\(self.serverPort)")
                self.listenForMessages()
                // Send an initial packet to get registered on the server
                self.sendMessage("Hello from Swift")
            case .failed(let error):
                print("PeerClient error starting: \(error)")
                self.connection = nil
            default:
                break
            }
        }
        
        connection?.start(queue: queue)
    }

    private func listenForMessages() {
        connection?.receiveMessage { [weak self] (content, context, isComplete, error) in
            if let content = content, !content.isEmpty {
                let message = String(data: content, encoding: .utf8) ?? ""
                print("Message received: \(message)")
                self?.onMessageReceived(message)
            }
            
            if error == nil {
                // Listen for the next message
                self?.listenForMessages()
            }
        }
    }

    func sendMessage(_ message: String) {
        guard let connection = connection else {
            print("Cannot send message, client is not running.")
            return
        }
        
        let data = message.data(using: .utf8)
        connection.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                print("Error sending message: \(error)")
            } else {
                print("Message sent: \(message)")
            }
        })
    }

    func stop() {
        guard let connection = connection else { return }
        connection.cancel()
        self.connection = nil
        print("PeerClient stopped.")
    }
}

File: peer_aos/app/src/main/java/dev/bluelemonade/peerapp/PeerClient.kt
package dev.bluelemonade.peerapp

import android.util.Log
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.util.concurrent.Executors

/**
 * A client for connecting to the RelayServer.
 *
 * This client handles sending and receiving UDP packets to and from the server.
 *
 * @param serverIp The IP address of the device running the RelayServer.
 * @param serverPort The port number the RelayServer is listening on (default is 8888).
 * @param onMessageReceived A callback function that is invoked with the message
 *                          received from another peer via the server.
 */
class PeerClient(
    private val serverIp: String,
    private val serverPort: Int = 8888,
    private val onMessageReceived: (String) -> Unit
) {
    private var socket: DatagramSocket? = null
    private val executor = Executors.newSingleThreadExecutor()
    private var isRunning = false

    /**
     * Starts the client.
     *
     * This initializes the socket, starts a listener thread for incoming messages,
     * and sends an initial packet to register with the server.
     */
    fun start() {
        if (isRunning) {
            Log.w("PeerClient", "Client is already running.")
            return
        }
        isRunning = true

        executor.execute {
            try {
                // Use a random available local port
                socket = DatagramSocket()
                val serverAddress = InetAddress.getByName(serverIp)
                Log.d("PeerClient", "Client started. Server: $serverIp:$serverPort")

                // Start a listener thread for messages from the server
                Thread { listenForMessages() }.start()

                // Send an initial packet to get registered on the server
                val initialMessage = "Hello from ${socket?.localPort}".toByteArray()
                val initialPacket = DatagramPacket(initialMessage, initialMessage.size, serverAddress, serverPort)
                socket?.send(initialPacket)

            } catch (e: Exception) {
                Log.e("PeerClient", "Error starting client", e)
                isRunning = false
            }
        }
    }

    private fun listenForMessages() {
        Log.d("PeerClient", "Listener thread started.")
        while (isRunning) {
            try {
                val buffer = ByteArray(2048) // Buffer for incoming data
                val packet = DatagramPacket(buffer, buffer.size)
                socket?.receive(packet) // This call blocks until a packet is received

                if (packet.length > 0) {
                    val message = String(packet.data, 0, packet.length)
                    Log.d("PeerClient", "Message received: $message")
                    onMessageReceived(message)
                }
            } catch (e: Exception) {
                if (isRunning) {
                    Log.e("PeerClient", "Error in listener thread", e)
                }
            }
        }
        Log.d("PeerClient", "Listener thread stopped.")
    }

    /**
     * Sends a message to the server, which will broadcast it to other peers.
     *
     * @param message The message to send. This can be any string, but for a real
     *                voice call app, you would send encoded audio data.
     */
    fun sendMessage(message: String) {
        if (!isRunning) {
            Log.w("PeerClient", "Cannot send message, client is not running.")
            return
        }
        executor.execute {
            try {
                val serverAddress = InetAddress.getByName(serverIp)
                val data = message.toByteArray()
                val packet = DatagramPacket(data, data.size, serverAddress, serverPort)
                socket?.send(packet)
                Log.d("PeerClient", "Message sent: $message")
            } catch (e: Exception) {
                Log.e("PeerClient", "Error sending message", e)
            }
        }
    }

    /**
     * Stops the client and closes the socket.
     */
    fun stop() {
        if (!isRunning) return
        isRunning = false
        executor.shutdown()
        socket?.close()
        socket = null
        Log.d("PeerClient", "Client stopped.")
    }
}